File: class26/python/GQS2601-GitHubActions_Workflow.py
#!/usr/bin/env python3
"""
GQS2601-GitHubActions_Workflow.py
==================================
Gerador e validador de workflow do GitHub Actions (conceitual).
Aula 26 — DevOps e GitHub Actions | Garantia da Qualidade de Software.

Uso:
    python3 GQS2601-GitHubActions_Workflow.py              # modo interativo
    python3 GQS2601-GitHubActions_Workflow.py --validate    # valida exemplos pré-definidos
"""

import sys
import re
from typing import Dict, List

# ──────────────────────────────────────────────
# 1. Validação de workflow YAML (simulada)
# ──────────────────────────────────────────────

REQUIRED_KEYS: Dict[str, List[str]] = {
    "name": ["É recomendado fornecer um nome para o workflow."],
    "on": ["O gatilho 'on' é obrigatório (push, pull_request, schedule, etc.)."],
    "jobs": ["Pelo menos um job deve ser definido."],
}

def validate_workflow(yaml_content: str) -> List[str]:
    """
    Valida conceitualmente um workflow YAML do GitHub Actions.

    Args:
        yaml_content: String contendo o YAML do workflow.

    Returns:
        Lista de problemas encontrados (vazia se tudo ok).
    """
    issues: List[str] = []

    # Verificações básicas com regex (não é um parser YAML completo)
    if not re.search(r'^name\s*:', yaml_content, re.MULTILINE):
        issues.append("[AVISO] 'name' não encontrado. " + REQUIRED_KEYS["name"][0])

    if not re.search(r'^on\s*:', yaml_content, re.MULTILINE):
        issues.append("[ERRO] 'on' não encontrado. " + REQUIRED_KEYS["on"][0])
    else:
        # Verificar se há pelo menos um evento
        if not re.search(r'push|pull_request|schedule|workflow_dispatch', yaml_content):
            issues.append("[AVISO] Nenhum evento conhecido encontrado em 'on'.")

    if not re.search(r'^jobs\s*:', yaml_content, re.MULTILINE):
        issues.append("[ERRO] 'jobs' não encontrado. " + REQUIRED_KEYS["jobs"][0])
    else:
        # Verificar se os jobs têm 'runs-on'
        job_matches = re.findall(r'^\s{2}(\w[\w-]*)\s*:', yaml_content, re.MULTILINE)
        if not job_matches:
            issues.append("[AVISO] Nenhum job nomeado encontrado em 'jobs'.")

        runs_on_count = len(re.findall(r'runs-on\s*:', yaml_content))
        if runs_on_count == 0:
            issues.append("[ERRO] Nenhum job possui 'runs-on'.")

        steps_count = len(re.findall(r'^\s{4}- name\s*:', yaml_content, re.MULTILINE)) + \
                      len(re.findall(r'^\s{4}- uses\s*:', yaml_content, re.MULTILINE))
        if steps_count == 0:
            issues.append("[AVISO] Nenhum step encontrado nos jobs (use 'steps:' com 'name' ou 'uses').")

    return issues


# ──────────────────────────────────────────────
# 2. Gerador de workflow template
# ──────────────────────────────────────────────

def generate_workflow_template(
    workflow_name: str = "CI - Build and Test",
    java_version: str = "17",
    python_version: str = "3.11",
    os_runner: str = "ubuntu-latest",
    use_matrix: bool = False,
) -> str:
    """
    Gera um template de workflow YAML do GitHub Actions.

    Args:
        workflow_name: Nome do workflow.
        java_version: Versão do Java (ex.: "17", "21").
        python_version: Versão do Python (ex.: "3.11").
        os_runner: Runner do GitHub Actions (ex.: "ubuntu-latest").
        use_matrix: Se True, gera com strategy.matrix para múltiplas versões.

    Returns:
        String com o YAML do workflow.
    """
    lines = [
        f"name: {workflow_name}",
        "",
        "on:",
        "  push:",
        "    branches: [ main ]",
        "  pull_request:",
        "    branches: [ main ]",
        "",
        "jobs:",
        "  test:",
        f"    runs-on: {os_runner}",
    ]

    if use_matrix:
        lines.extend([
            "",
            "    strategy:",
            "      matrix:",
            f"        java: [17, 21]",
            f"        os: [ubuntu-latest, windows-latest]",
            "",
            "    steps:",
            "      - name: Checkout code",
            "        uses: actions/checkout@v4",
            "",
            "      - name: Set up JDK",
            "        uses: actions/setup-java@v4",
            "        with:",
            "          java-version: ${{ matrix.java }}",
            "          distribution: 'temurin'",
            "",
            "      - name: Compile and test with Maven",
            "        run: mvn clean test",
        ])
    else:
        lines.extend([
            "",
            "    steps:",
            "      - name: Checkout code",
            "        uses: actions/checkout@v4",
            "",
            "      - name: Set up JDK",
            "        uses: actions/setup-java@v4",
            "        with:",
            f"          java-version: '{java_version}'",
            "          distribution: 'temurin'",
            "",
            "      - name: Cache Maven dependencies",
            "        uses: actions/cache@v4",
            "        with:",
            "          path: ~/.m2/repository",
            "          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}",
            "          restore-keys: |",
            "            ${{ runner.os }}-maven-",
            "",
            "      - name: Run tests with coverage",
            "        run: mvn clean test jacoco:report",
            "",
            "      - name: Upload coverage report",
            "        uses: actions/upload-artifact@v4",
            "        with:",
            "          name: coverage-report",
            "          path: target/site/jacoco/",
        ])

    return "\n".join(lines) + "\n"


# ──────────────────────────────────────────────
# 3. Exemplos pré-definidos para validação
# ──────────────────────────────────────────────

WORKFLOW_EXAMPLE_VALID = """name: CI - Java 17
on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup Java
        uses: actions/setup-java@v4
        with:
          java-version: '17'
          distribution: 'temurin'
      - name: Run tests
        run: mvn test
"""

WORKFLOW_EXAMPLE_INVALID = """name:
on:
  push

  test:
    steps:
      - run: echo hello
"""


# ──────────────────────────────────────────────
# 4. Modo interativo
# ──────────────────────────────────────────────

def interactive_mode() -> None:
    """Executa o script em modo interativo."""
    print("=" * 60)
    print("  GQS2601 - Gerador de Workflow GitHub Actions")
    print("  Garantia da Qualidade de Software — Aula 26")
    print("=" * 60)

    while True:
        print("\nOpções:")
        print("  1 - Validar workflow YAML existente")
        print("  2 - Gerar template de workflow")
        print("  3 - Validar exemplos pré-definidos")
        print("  0 - Sair")

        choice = input("\nEscolha: ").strip()

        if choice == "1":
            print("\nCole o YAML do workflow (digite 'EOF' em linha vazia para finalizar):")
            lines = []
            while True:
                line = input()
                if line.strip().upper() == "EOF":
                    break
                lines.append(line)
            yaml_text = "\n".join(lines)
            issues = validate_workflow(yaml_text)
            if issues:
                print("\n❌ Problemas encontrados:")
                for issue in issues:
                    print(f"   • {issue}")
            else:
                print("\n✅ Workflow válido! Nenhum problema encontrado.")

        elif choice == "2":
            print("\n--- Gerar Template de Workflow ---")
            name = input("Nome do workflow [CI - Build and Test]: ").strip() or "CI - Build and Test"
            java_ver = input("Versão Java [17]: ").strip() or "17"
            py_ver = input("Versão Python [3.11]: ").strip() or "3.11"
            os_runner = input("Runner OS [ubuntu-latest]: ").strip() or "ubuntu-latest"
            matrix_str = input("Usar matrix (s/N)? ").strip().lower()
            use_matrix = matrix_str in ("s", "sim", "y", "yes")

            print("\n--- Workflow Gerado ---\n")
            print(generate_workflow_template(name, java_ver, py_ver, os_runner, use_matrix))

        elif choice == "3":
            print("\n--- Validando workflow VÁLIDO ---")
            issues = validate_workflow(WORKFLOW_EXAMPLE_VALID)
            if issues:
                for i in issues:
                    print(f"   • {i}")
            else:
                print("   ✅ Válido!")

            print("\n--- Validando workflow INVÁLIDO ---")
            issues = validate_workflow(WORKFLOW_EXAMPLE_INVALID)
            if issues:
                for i in issues:
                    print(f"   • {i}")
            else:
                print("   ✅ Válido!")

        elif choice == "0":
            print("\nEncerrando. Até a próxima aula!")
            break

        else:
            print("\nOpção inválida. Tente novamente.")


# ──────────────────────────────────────────────
# 5. Ponto de entrada
# ──────────────────────────────────────────────

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "--validate":
        print("=" * 60)
        print("  GQS2601 - Validação de Workflows")
        print("=" * 60)

        print("\n1. Workflow válido:")
        issues_valid = validate_workflow(WORKFLOW_EXAMPLE_VALID)
        if issues_valid:
            for i in issues_valid:
                print(f"   ❌ {i}")
        else:
            print("   ✅ Nenhum problema.")

        print("\n2. Workflow inválido:")
        issues_invalid = validate_workflow(WORKFLOW_EXAMPLE_INVALID)
        if issues_invalid:
            for i in issues_invalid:
                print(f"   ❌ {i}")
        else:
            print("   ✅ Nenhum problema.")

        print("\n3. Template gerado (com matrix):")
        print(generate_workflow_template(use_matrix=True))

        sys.exit(0)

    interactive_mode()

File: class26/java/GQS2601_GitHubActions_Workflow.java
/**
 * GQS2601-GitHubActions_Workflow.java
 * ======================================
 * Gerador e validador conceitual de workflow do GitHub Actions.
 * Aula 26 — DevOps e GitHub Actions | Garantia da Qualidade de Software.
 *
 * Compilar:  javac GQS2601-GitHubActions_Workflow.java
 * Executar:  java GQS2601-GitHubActions_Workflow
 */

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class GQS2601_GitHubActions_Workflow {

    // ──────────────────────────────────────────────
    // 1. Validação de workflow YAML (simulada)
    // ──────────────────────────────────────────────

    /**
     * Valida conceitualmente um workflow YAML do GitHub Actions.
     * @param yamlContent String contendo o YAML do workflow.
     * @return Lista de problemas encontrados (vazia se tudo ok).
     */
    public static List<String> validateWorkflow(String yamlContent) {
        List<String> issues = new ArrayList<>();

        // Verifica 'name'
        if (!yamlContent.contains("name:")) {
            issues.add("[AVISO] 'name' não encontrado. Recomenda-se fornecer um nome.");
        }

        // Verifica 'on'
        if (!yamlContent.contains("on:")) {
            issues.add("[ERRO] 'on' não encontrado. O gatilho é obrigatório.");
        } else {
            Pattern eventPattern = Pattern.compile("push|pull_request|schedule|workflow_dispatch");
            Matcher matcher = eventPattern.matcher(yamlContent);
            if (!matcher.find()) {
                issues.add("[AVISO] Nenhum evento conhecido encontrado em 'on'.");
            }
        }

        // Verifica 'jobs'
        if (!yamlContent.contains("jobs:")) {
            issues.add("[ERRO] 'jobs' não encontrado. Pelo menos um job deve ser definido.");
        } else {
            // Verifica 'runs-on'
            if (!yamlContent.contains("runs-on:")) {
                issues.add("[ERRO] Nenhum job possui 'runs-on'.");
            }
            // Verifica steps
            if (!yamlContent.contains("steps:")) {
                issues.add("[AVISO] Nenhum step encontrado. Use 'steps:' com ações.");
            }
        }

        return issues;
    }

    // ──────────────────────────────────────────────
    // 2. Gerador de template de workflow
    // ──────────────────────────────────────────────

    public static String generateWorkflowTemplate(
            String workflowName,
            String javaVersion,
            String osRunner,
            boolean useMatrix) {

        StringBuilder sb = new StringBuilder();

        sb.append("name: ").append(workflowName).append("\n");
        sb.append("\n");
        sb.append("on:\n");
        sb.append("  push:\n");
        sb.append("    branches: [ main ]\n");
        sb.append("  pull_request:\n");
        sb.append("    branches: [ main ]\n");
        sb.append("\n");
        sb.append("jobs:\n");
        sb.append("  test:\n");
        sb.append("    runs-on: ").append(osRunner).append("\n");

        if (useMatrix) {
            sb.append("\n");
            sb.append("    strategy:\n");
            sb.append("      matrix:\n");
            sb.append("        java: [17, 21]\n");
            sb.append("        os: [ubuntu-latest, windows-latest]\n");
            sb.append("\n");
            sb.append("    steps:\n");
            sb.append("      - name: Checkout code\n");
            sb.append("        uses: actions/checkout@v4\n");
            sb.append("\n");
            sb.append("      - name: Set up JDK\n");
            sb.append("        uses: actions/setup-java@v4\n");
            sb.append("        with:\n");
            sb.append("          java-version: ${{ matrix.java }}\n");
            sb.append("          distribution: 'temurin'\n");
            sb.append("\n");
            sb.append("      - name: Compile and test with Maven\n");
            sb.append("        run: mvn clean test\n");
        } else {
            sb.append("\n");
            sb.append("    steps:\n");
            sb.append("      - name: Checkout code\n");
            sb.append("        uses: actions/checkout@v4\n");
            sb.append("\n");
            sb.append("      - name: Set up JDK\n");
            sb.append("        uses: actions/setup-java@v4\n");
            sb.append("        with:\n");
            sb.append("          java-version: '").append(javaVersion).append("'\n");
            sb.append("          distribution: 'temurin'\n");
            sb.append("\n");
            sb.append("      - name: Cache Maven dependencies\n");
            sb.append("        uses: actions/cache@v4\n");
            sb.append("        with:\n");
            sb.append("          path: ~/.m2/repository\n");
            sb.append("          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}\n");
            sb.append("          restore-keys: |\n");
            sb.append("            ${{ runner.os }}-maven-\n");
            sb.append("\n");
            sb.append("      - name: Run tests with coverage\n");
            sb.append("        run: mvn clean test jacoco:report\n");
            sb.append("\n");
            sb.append("      - name: Upload coverage report\n");
            sb.append("        uses: actions/upload-artifact@v4\n");
            sb.append("        with:\n");
            sb.append("          name: coverage-report\n");
            sb.append("          path: target/site/jacoco/\n");
        }

        return sb.toString();
    }

    // ──────────────────────────────────────────────
    // 3. Exemplos pré-definidos
    // ──────────────────────────────────────────────

    private static final String WORKFLOW_VALID =
            "name: CI - Java 17\n" +
            "on:\n" +
            "  push:\n" +
            "    branches: [ main ]\n" +
            "  pull_request:\n" +
            "    branches: [ main ]\n" +
            "\n" +
            "jobs:\n" +
            "  test:\n" +
            "    runs-on: ubuntu-latest\n" +
            "    steps:\n" +
            "      - name: Checkout\n" +
            "        uses: actions/checkout@v4\n" +
            "      - name: Setup Java\n" +
            "        uses: actions/setup-java@v4\n" +
            "        with:\n" +
            "          java-version: '17'\n" +
            "          distribution: 'temurin'\n" +
            "      - name: Run tests\n" +
            "        run: mvn test\n";

    private static final String WORKFLOW_INVALID =
            "name:\n" +
            "on:\n" +
            "  push\n" +
            "\n" +
            "  test:\n" +
            "    steps:\n" +
            "      - run: echo hello\n";

    // ──────────────────────────────────────────────
    // 4. Modo interativo
    // ──────────────────────────────────────────────

    public static void interactiveMode(Scanner scanner) {
        System.out.println("=".repeat(60));
        System.out.println("  GQS2601 - Gerador de Workflow GitHub Actions");
        System.out.println("  Garantia da Qualidade de Software — Aula 26");
        System.out.println("=".repeat(60));

        while (true) {
            System.out.println("\nOpções:");
            System.out.println("  1 - Validar workflow YAML existente");
            System.out.println("  2 - Gerar template de workflow");
            System.out.println("  3 - Validar exemplos pré-definidos");
            System.out.println("  0 - Sair");

            System.out.print("\nEscolha: ");
            String choice = scanner.nextLine().trim();

            switch (choice) {
                case "1":
                    System.out.println("\nCole o YAML do workflow (digite 'EOF' em linha vazia para finalizar):");
                    StringBuilder yamlInput = new StringBuilder();
                    while (scanner.hasNextLine()) {
                        String line = scanner.nextLine();
                        if (line.trim().equalsIgnoreCase("EOF")) break;
                        yamlInput.append(line).append("\n");
                    }
                    List<String> issues = validateWorkflow(yamlInput.toString());
                    if (issues.isEmpty()) {
                        System.out.println("\n✅ Workflow válido! Nenhum problema encontrado.");
                    } else {
                        System.out.println("\n❌ Problemas encontrados:");
                        for (String issue : issues) {
                            System.out.println("   • " + issue);
                        }
                    }
                    break;

                case "2":
                    System.out.println("\n--- Gerar Template de Workflow ---");
                    System.out.print("Nome do workflow [CI - Build and Test]: ");
                    String name = scanner.nextLine().trim();
                    if (name.isEmpty()) name = "CI - Build and Test";

                    System.out.print("Versão Java [17]: ");
                    String javaVer = scanner.nextLine().trim();
                    if (javaVer.isEmpty()) javaVer = "17";

                    System.out.print("Runner OS [ubuntu-latest]: ");
                    String os = scanner.nextLine().trim();
                    if (os.isEmpty()) os = "ubuntu-latest";

                    System.out.print("Usar matrix (s/N)? ");
                    String matrixChoice = scanner.nextLine().trim().toLowerCase();
                    boolean useMatrix = matrixChoice.equals("s") || matrixChoice.equals("sim");

                    System.out.println("\n--- Workflow Gerado ---\n");
                    System.out.println(generateWorkflowTemplate(name, javaVer, os, useMatrix));
                    break;

                case "3":
                    System.out.println("\n--- Validando workflow VÁLIDO ---");
                    List<String> validIssues = validateWorkflow(WORKFLOW_VALID);
                    if (validIssues.isEmpty()) {
                        System.out.println("   ✅ Válido!");
                    } else {
                        validIssues.forEach(i -> System.out.println("   ❌ " + i));
                    }

                    System.out.println("\n--- Validando workflow INVÁLIDO ---");
                    List<String> invalidIssues = validateWorkflow(WORKFLOW_INVALID);
                    if (invalidIssues.isEmpty()) {
                        System.out.println("   ✅ Válido!");
                    } else {
                        invalidIssues.forEach(i -> System.out.println("   ❌ " + i));
                    }

                    System.out.println("\n--- Template gerado (com matrix) ---");
                    System.out.println(generateWorkflowTemplate("CI - Matrix Build", "17", "ubuntu-latest", true));
                    break;

                case "0":
                    System.out.println("\nEncerrando. Até a próxima aula!");
                    return;

                default:
                    System.out.println("\nOpção inválida. Tente novamente.");
            }
        }
    }

    // ──────────────────────────────────────────────
    // 5. Ponto de entrada
    // ──────────────────────────────────────────────

    public static void main(String[] args) {
        if (args.length > 0 && args[0].equals("--validate")) {
            System.out.println("=".repeat(60));
            System.out.println("  GQS2601 - Validação de Workflows");
            System.out.println("=".repeat(60));

            System.out.println("\n1. Workflow válido:");
            List<String> issuesValid = validateWorkflow(WORKFLOW_VALID);
            if (issuesValid.isEmpty()) {
                System.out.println("   ✅ Nenhum problema.");
            } else {
                issuesValid.forEach(i -> System.out.println("   ❌ " + i));
            }

            System.out.println("\n2. Workflow inválido:");
            List<String> issuesInvalid = validateWorkflow(WORKFLOW_INVALID);
            if (issuesInvalid.isEmpty()) {
                System.out.println("   ✅ Nenhum problema.");
            } else {
                issuesInvalid.forEach(i -> System.out.println("   ❌ " + i));
            }

            System.out.println("\n3. Template gerado (com matrix):");
            System.out.println(generateWorkflowTemplate("CI - Matrix Build", "17", "ubuntu-latest", true));

            System.exit(0);
        }

        try (Scanner scanner = new Scanner(System.in)) {
            interactiveMode(scanner);
        }
    }
}
